### PR TITLE
 Rewrite the library to improve safety and prevent extra allocations.

### PR DESCRIPTION
**Pull request by emilio**
#### Disclaimer

Don't feel obligated to merge this, I'm aware that is a fairly radical change to the library. I was reading over it as I said in https://github.com/servo/servo/pull/12186 and couldn't even manage it to get to compile (see below).

I also noticed quite a few issues listed below, and noticed that a few things could be done with fewer allocations and FFI types.

I've added some tests to make sure it fits for Servo (basically the `partial_file` check).

That being said, you're the owner of this repo, so you're the one who has to review my code and either accept the changes or reject them.

---

This is a (major) rewrite of the library to avoid extra complexity, avoid
overriding symbols, and a bunch of other stuff.

I've verified that no leaks happen on the happy paths and the error paths 
I've been able to test, and tried to reduce the logic to handle errors to a 
point where is reviewable easily, but of course it's C so it needs careful 
review anyway.

Some of the problems with the previous implementation were:
- Too generic public symbols (could conflict easily).
- Leaks in error paths and loops.
- Hackish ways to get durations, reporting invalid ones.
- Lazy non-thread-safe initialization.
- Didn't work with newer versions of LibAV (see error below).

```
ffi/info.c: In function ‘iterate_over_streams’:
ffi/info.c:72:9: warning: implicit declaration of function ‘avcodec_parameters_to_context’ [-Wimplicit-function-declaration]
     if (avcodec_parameters_to_context(avctx, ic->streams[i]->codecpar) < 0) {
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ffi/info.c:72:60: error: ‘AVStream {aka struct AVStream}’ has no member named ‘codecpar’; did you mean ‘codec’?
     if (avcodec_parameters_to_context(avctx, ic->streams[i]->codecpar) < 0) {
```

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -1,16 +1,46 @@
 use std::time::Duration;
 use KnownTypes;
+use video_metadata::{av_strerror_safe, vmrs_result};
 
 #[derive(Clone, Debug, PartialEq)]
 pub struct Size {
-    pub width: u16,
-    pub height: u16,
+    pub width: u32,
```
Comment (GuillaumeGomez) [inline src/types.rs:7]: Isn't 65536 enough of a size?

File: src/video_metadata.rs [hunk 2]
```diff
@@ -1,154 +1,257 @@
...
+                                out: *mut vmrs_metadata) -> c_int;
+
+    fn vmrs_read_info_from_buffer(buffer: *const u8,
+                                  size: u32,
+                                  out: *mut vmrs_metadata) -> c_int;
+
+    fn vmrs_metadata_free(metadata: *mut vmrs_metadata);
 }
 
+const AV_TIME_BASE: u64 = 1000000;
```
Comment (GuillaumeGomez) [inline src/video_metadata.rs:33]: Why write this in here? If it changes on C side, we'll have to update it directly and I don't find this very convenient. :-/

File: src/video_metadata.rs [hunk 3]
```diff
@@ -1,154 +1,257 @@
...
 
+// TODO: use rust-bindgen for this, or better yet, create/use libav bindings.
 extern "C" {
-    fn get_information(filename: *const c_char) -> CRet;
-    fn free_ret(r: *mut CRet);
-    fn get_time_base() -> i64;
-    fn get_information_from_buffer(buffer: *const c_char, size: c_uint) -> CRet;
+    fn av_strerror(errnum: c_int, errbuf: *mut c_char, errbuf_size: usize) -> c_int;
+
+    fn vmrs_initialize();
```
Comment (GuillaumeGomez) [inline src/video_metadata.rs:16]: Isn't it better to keep it on C side only?

File: src/types.rs [hunk 4]
```diff
@@ -1,16 +1,46 @@
...
+            "Alloc failure"
+        } else if self.0 == vmrs_result::VMRS_FORMAT_NOT_AVAILABLE as i32 {
+            "Format wasn't available"
+        } else {
+            "vmrs error not handled, this is a bug"
+        }.to_owned()
+    }
+}
+
+
```
Comment (GuillaumeGomez) [inline src/types.rs]: Extra empty line.

File: ffi/vmrs.c [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    int video_stream_index;
+    int audio_stream_index;
+
+    ret = avformat_open_input(&format_ctx, filename, NULL, NULL);
+    if (ret < 0) {
+        // NB: avformat_open_input already frees the context on failure, though
+        // doesn't free io_context if it's a custom one.
+        format_ctx = NULL;
+        // NB: We return the `ret` value, which is the real libav format.
+        goto errorexit;
```
Comment (GuillaumeGomez) [inline ffi/vmrs.c:151]: I'm really not a big fan of `goto` calls but in this case it doesn't make the code harder to read. Let's say it's ok then. :wink:

**Comment by GuillaumeGomez** [comment]
Thanks so much for this PR! Don't forget to add yourself into Cargo's author list. Just one thing: could you keep a little C test code somewhere please? I find it very useful to test memory leaks and stuff (`valgrind` and `cargo test` aren't very friendly).

File: src/video_metadata.rs (hunk 3 above)
Comment (emilio) [inline src/video_metadata.rs:16]: Yes, but doing it thread-safe would require me to `#ifdef` posix once's or windows', and that's definitely not better :)

File: src/video_metadata.rs (hunk 2 above)
Comment (emilio) [inline src/video_metadata.rs:33]: While that's a fair concern, I don't expect LibAV to change their time base all of a sudden. Ideally we'd generate bindings for it, but this is the approach taken at let's say, libc. What if any syscall number changes?

File: src/types.rs (hunk 1 above)
Comment (emilio) [inline src/types.rs:7]: It probably is, but I think it's worth to stick with at least the size libav provides. Can definitely do this change though.

File: ffi/vmrs.c (hunk 5 above)
Comment (emilio) [inline ffi/vmrs.c:151]: Yeah, I'm not a big fan of it either, but it simplifies a lot of code that needs careful cleanup :P

File: src/video_metadata.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/video_metadata.rs:33]: As far as I know, you add syscall but you don't change their number. The only case it happened was between 32 and 64 bits.

Comment (emilio) [inline src/video_metadata.rs:33]: Yes, that was the point I was trying to make, that binary interfaces don't use to change in shared libraries for this very reason.

Comment (GuillaumeGomez) [inline src/video_metadata.rs:33]: Well, I still think this isn't a good idea but I won't bother you with it. :)

File: src/types.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/types.rs:7]: I don't mind either way. Let's say we schedule to handle very big videos. :stuck_out_tongue:

**Comment by GuillaumeGomez** [comment]
Before I merge, can you clean up a bit your commit history please? :)

**Comment by emilio** [comment]
All should be set! :)

**Comment by GuillaumeGomez** [comment]
That's just great! Thanks so much for this!

**Pull request merged**